drivers/sdhc/emmc_host.c

```diff
@@ -0,0 +1,1337 @@
+	if (read) {
+		ret = read_data_port(emmc, data);
+	} else {
+		ret = write_data_port(emmc, data);
+	}
+#else
+	ret = wait_xfr_complete(emmc, data->timeout_ms);
+#endif
+
+#if !defined(CONFIG_EMMC_HOST_AUTO_STOP)
+	emmc_stop_transfer(emmc);
+#endif
+	return ret;
+}
+
+static int emmc_request(const struct device *dev, struct sdhc_command *cmd, struct sdhc_data *data)
+{
+	struct emmc_data *emmc = dev->data;
+	int ret;
+
+	LOG_DBG("");
+
+	if (data) {
+		switch (cmd->opcode) {
+		case SD_WRITE_SINGLE_BLOCK:
+		case SD_WRITE_MULTIPLE_BLOCK:
+			LOG_DBG("SD_WRITE_SINGLE_BLOCK");
+			ret = emmc_xfr(emmc, cmd, data, false);
+			break;
+
+		case SD_READ_SINGLE_BLOCK:
+		case SD_READ_MULTIPLE_BLOCK:
+			LOG_DBG("SD_READ_SINGLE_BLOCK");
+			ret = emmc_xfr(emmc, cmd, data, true);
+			break;
+
+		case MMC_SEND_EXT_CSD:
+			LOG_DBG("EMMC_HOST_SEND_EXT_CSD");
+			ret = emmc_send_cmd_data(emmc, MMC_SEND_EXT_CSD, cmd, data, true);
+			break;
+
+		default:
+			ret = emmc_send_cmd_data(emmc, cmd->opcode, cmd, data, true);
+		}
+	} else {
+		ret = emmc_send_cmd_no_data(emmc, cmd->opcode, cmd);
+
+		/*fixme: WA for adding delay after card switch to HS200 mode */
+		if (cmd->opcode == SD_SEND_STATUS) {
+			k_msleep(10);
+		}
+	}
+
+	return ret;
+}
+
+static int emmc_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	struct emmc_data *emmc = dev->data;
+	volatile struct emmc_reg *regs = emmc->regs;
+	struct sdhc_io *host_io = &emmc->host_io;
+	int ret;
+
+	LOG_DBG("emmc I/O: DW %d, Clk %d Hz, card power state %s, voltage %s", ios->bus_width,
+		ios->clock, ios->power_mode == SDHC_POWER_ON ? "ON" : "OFF",
+		ios->signal_voltage == SD_VOL_1_8_V ? "1.8V" : "3.3V");
+
+	if (ios->clock && (ios->clock > emmc->props.f_max || ios->clock < emmc->props.f_min)) {
+		LOG_ERR("Invalid argument for clock freq: %d Support max:%d and Min:%d", ios->clock,
+			emmc->props.f_max, emmc->props.f_min);
+		return -EINVAL;
+	}
+
+	/* Set HC clock */
+	if (host_io->clock != ios->clock) {
+		LOG_DBG("Clock: %d", host_io->clock);
+		if (ios->clock != 0) {
+			/* Enable clock */
+			LOG_DBG("CLOCK: %d", ios->clock);
+			if (!emmc_clock_set(emmc, ios->clock)) {
+				return -ENOTSUP;
+			}
+		}
+		host_io->clock = ios->clock;
+	}
+
+	/* Set data width */
+	if (host_io->bus_width != ios->bus_width) {
+		LOG_DBG("bus_width: %d", host_io->bus_width);
+		SET_BITS(regs->host_ctrl1, EMMC_HOST_CTRL1_EXT_DAT_WIDTH_LOC,
+			 EMMC_HOST_CTRL1_EXT_DAT_WIDTH_MASK,
+			 ios->bus_width == SDHC_BUS_WIDTH8BIT ? 1 : 0);
+
+		regs->host_ctrl1 |= ios->bus_width << EMMC_HOST_CTRL1_DAT_WIDTH_LOC;
+		host_io->bus_width = ios->bus_width;
+	}
+
+	/* Set HC signal voltage */
+	if (ios->signal_voltage != host_io->signal_voltage) {
+		LOG_DBG("signal_voltage: %d", ios->signal_voltage);
+		ret = emmc_set_voltage(emmc, ios->signal_voltage);
+		if (ret) {
+			LOG_ERR("Set signal volatge failed failed:%d", ret);
+			return ret;
+		}
+		host_io->signal_voltage = ios->signal_voltage;
+		k_msleep(10u);
```

Comment:
What is the reason for the delay here? I assume there is some requirement in the datasheet, but would like a comment to that effect here

Reply:
This delay added as a safe point since power switching happens (not a requirement as per HC spec , section 3.3 SD Bus Power control). I removed this delay to avoid unnecessary confusion.